core/trino-main/src/main/java/io/trino/connector/system/TableCommentSystemTable.java

```diff
@@ -100,60 +97,83 @@ public RecordCursor cursor(ConnectorTransactionHandle transactionHandle, Connect
         for (String catalog : listCatalogNames(session, metadata, accessControl, catalogFilter)) {
             QualifiedTablePrefix prefix = tablePrefix(catalog, schemaFilter, tableFilter);
 
-            Set<SchemaTableName> names = ImmutableSet.of();
-            Map<SchemaTableName, ViewInfo> views = ImmutableMap.of();
-            Map<SchemaTableName, ViewInfo> materializedViews = ImmutableMap.of();
-            try {
-                materializedViews = getMaterializedViews(session, metadata, accessControl, prefix);
-                views = getViews(session, metadata, accessControl, prefix);
-                // Some connectors like blackhole, accumulo and raptor don't return views in listTables
-                // Materialized views are consistently returned in listTables by the relevant connectors
-                names = union(listTables(session, metadata, accessControl, prefix), views.keySet());
-            }
-            catch (TrinoException e) {
-                // listTables throws an exception if cannot connect the database
-                LOG.warn(e, "Failed to get tables for catalog: %s", catalog);
-            }
-
-            for (SchemaTableName name : names) {
-                Optional<String> comment = Optional.empty();
+            if (prefix.getTableName().isPresent()) {
+                QualifiedObjectName relationName = new QualifiedObjectName(catalog, prefix.getSchemaName().orElseThrow(), prefix.getTableName().get());
+                RelationComment relationComment;
                 try {
-                    comment = getComment(session, prefix, name, views, materializedViews);
+                    relationComment = getRelationComment(session, relationName);
                 }
                 catch (RuntimeException e) {
-                    // getTableHandle may throw an exception (e.g. Cassandra connector doesn't allow case insensitive column names)
-                    LOG.warn(e, "Failed to get metadata for table: %s", name);
+                    LOG.warn(e, "Failed to get comment for relation: %s", relationName);
+                    relationComment = new RelationComment(false, Optional.empty());
+                }
+                if (relationComment.found()) {
+                    SchemaTableName schemaTableName = relationName.asSchemaTableName();
+                    // Consulting accessControl first would be simpler but some AccessControl implementations may have issues when asked for a relation that does not exist.
+                    if (accessControl.filterTables(session.toSecurityContext(), catalog, ImmutableSet.of(schemaTableName)).contains(schemaTableName)) {
+                        table.addRow(catalog, schemaTableName.getSchemaName(), schemaTableName.getTableName(), relationComment.comment().orElse(null));
+                    }
+                }
+            }
+            else {
+                List<RelationCommentMetadata> relationComments = metadata.listRelationComments(
```

Comment:
Should probably run access control checks again here in case the connector does not behave well in filtering out tables that should not be accessible. 

I also saw there's an access control `filterColumns` method, do we need to be applying that here?

Reply:
> I also saw there's an access control filterColumns method, do we need to be applying that here?

i don't think it's applicable.

> Should probably run access control checks again here in case the connector does not behave well in filtering out tables that should not be accessible.

if we want that, we would need to let the connector declare whether AC was consulted or not. otherwise we would be calling AC twice, which is redundant, and considerable cost (in case of many relations).

i had this in this PR, see before `Fixup simplify streamRelationComments: require relationFilter to be applied` commit. i didn't like the additional complexity though.

Note also that this is a matter of defining contract. We do not expect connectors to ignore `schemaName` parameter (`tablePrefix` in other methods). So we can choose not to expect connectors to ignore `relationFilters`.